// genericmodel/lib/genericmodel.js
var mysql = require('mysql');

var GenericModel = function() {
	
  	var model = {
		
		query: function(pool){
			
			var query = {	
			
				execute: function(sql,fnc){	
					pool.getConnection(function(err, connection) {
						connection.query(sql, function(err, result) {  
		    				connection.end();
							fnc(result,err);
						});
					});		
				}
				
			}
			return query;	
		},
		
		crud: function(pool,table){
			
			var crud = {
			
				findAll: function(fnc){
					pool.getConnection(function(err, connection) {			  
						connection.query('SELECT * FROM '+table, function(err, rows) {  
							//var contatos = [];
							//rows.forEach(function(item){
							//	contatos.push(item);
							//});
				    		connection.end();
							//fnc(contatos);
							fnc(rows);
						});
					});	
				},
			
				findById: function(id,fnc){
					pool.getConnection(function(err, connection) {			  
						connection.query('SELECT * FROM '+table+' WHERE id='+id, function(err, rows) {  
				    		connection.end();
							fnc(rows[0]);
						});
					});	
				},
			
				insert: function(param,fnc){
					pool.getConnection(function(err, connection) {			  
						connection.query('INSERT INTO '+table+' SET ?',param, function(err, result) {  
				    		connection.end();
							fnc(result);
						});
					});	
				},
			
				update: function(param,fnc){
					pool.getConnection(function(err, connection) {			  
						connection.query('UPDATE '+table+' SET ? WHERE id='+param.id,param, function(err, result) {  
				    		connection.end();
							fnc(result);
						});
					});	
				},
			
				delete: function(id,fnc){
					pool.getConnection(function(err, connection) {
						connection.query('DELETE FROM '+table+' WHERE id='+id,function(err, result) {  						
				    		connection.end();
							fnc(result);
						});
					});	
				}
			}
			return crud;
		}
		
	}
	
	return model;
}

module.exports = function() {
	return new GenericModel();
}

module.exports.getPool = function(props) {
	return mysql.createPool(props);
}
